File: solutions/haskell/luhn/1/src/Luhn.hs
module Luhn (isValid) where

import Data.Char

isValid' :: [Int] -> Bool
isValid' n = (
  sum . 
  map (\v -> if v > 9 then v - 9 else v) $ 
    map (\(b, v) -> if b then 2 * v else v) $ 
      (zip (cycle [False, True]) (reverse n))) `rem` 10 == 0


isValid :: String -> Bool
isValid n = 
  let digits = (map (\x -> (ord x - ord '0')) $ filter isDigit n)
  in length digits > 1 && isValid' digits

File: solutions/python/luhn/1/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num.replace(" ", "")

    def valid(self):
        if len(self.card_num) <= 1 or not self.card_num.isdigit():
            return False

        digits = [int(c) for c in self.card_num]

        for i in range(len(digits) - 2, -1, -2):
            doubled = digits[i] * 2
            digits[i] = doubled - 9 if doubled > 9 else doubled

        return sum(digits) % 10 == 0
